Add Process Information Extension (Disabled by Default)

time-killer-games opened the pull request:
Yeah...

If someone ever hands you a project file with this extension being used; it is disabled by default, so please don't add a fucking retarded prompt until AFTER we sort out the IDE checkbox to optionally disable it. Actually, no - don't add it at all. I'll do it myself the correct way once we have that checkbox, the correct way involves not using a "confirmed_idiot" variable or an over-exaggerated scary message.

fundies commented:
Disabled by default won't have much meaning in egm but do as you wish. I'm not working on enigma anymore so feel free to muck up everything while I'm gone

time-killer-games commented:
@RobertBColton the windows-related code is ready for your review. Also I'd like some clarification on what fundies means exactly by it being disabled by default 'not effecting egm'.

RobertBColton commented:
EGM saves enabled extensions so you don't always have to toggle them.

time-killer-games commented:
@RobertBColton is there any good reason why the local IDE can't have this saved in it's settings and not embedded in the project itself? Also, just a guess, it was fundies' idea to save it in the project files, wasn't it? That would explain a lot, because that would be an effort to prevent me from making extensions like this.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Win32/procinfo.cpp:
```diff
@@ -0,0 +1,283 @@
+using std::vector;
+using std::size_t;
+
+static inline wstring widen(string str) {
+  size_t wchar_count = str.size() + 1;
+  vector<wchar_t> buf(wchar_count);
+  return wstring{ buf.data(), (size_t)MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, buf.data(), (int)wchar_count) };
+}
+
+static inline string narrow(wstring wstr) {
+  int nbytes = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), (int)wstr.length(), NULL, 0, NULL, NULL);
+  vector<char> buf(nbytes);
+  return string{ buf.data(), (size_t)WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), (int)wstr.length(), buf.data(), nbytes, NULL, NULL) };
```
Don't we have these in `shared/` yet? If not, the ones in the engine should be moved there and your extension should link them.

time-killer-games replied:
The reason I don't do that currently is because I want these files to be self-contained, meaning they can build independently apart from the rest of the engine - this extension for enigma I am also building as a dll for a gamemaker extension, and this makes that easier to maintain. But if you guys don't care about making my life easier, that's ok, just confirm you still want me to change it and I'll go right ahead and do it. lel

RobertBColton replied:
Well the way I see it is you are bordering between an official ENIGMA extension and a regular user extension. I am not so convinced this needs to be in the main ENIGMA, since none of it is actual GM API? I am also not completely convinced of the necessity for its use case either. Regardless, ideally, all main official default included ENIGMA extensions would share code in a modular way. If maintainability concerns you then, honestly, just distributing this on the forum or etc as an unofficial user extension would be appropriate.

time-killer-games replied:
Well, ENIGMA by definition is more than a GameMaker clone, but rather, an augmentation, so it would make sense if we had stuff in our engine that GameMaker doesn't, which is why I thought this extension might be acceptable. I'll go ahead and move those functions to shared like you wanted in a different pr, and once that gets merged we merge those changes into this branch and I'll include the functions from the shared header.

Since the last time I've talked with Josh, you, him, and fundies seem to all be on the same page that this doesn't belong in the engine. I'll close this pr if I can confirm from Josh he still feels this way, and I won't be surprised if he hasn't, as nothing has changed that would influence his opinion, I'm just reluctant to seeing all of this hard work I put into it not being an official part of ENIGMA. I can't get everything I want in life, jaja...

RobertBColton replied:
I definitely see that point of view, that ENIGMA certainly can have features that are not in GM, but that doesn't mean the whole world as you already see. We have to be careful about going too far overboard with features that aren't really going to be used that essentially, lest we make the engine too hard to maintain.

I wasn't necessarily decided on the usefulness of it though either, so I wasn't saying this isn't a useful or essential extension. Just see what Josh thinks about what I said.

time-killer-games replied:
You guys are just being honest, and I can't fault you for that. Will do.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/procinfo.h:
```diff
@@ -0,0 +1,142 @@
+
+#ifdef _WiN32
+#define byte __windows_byte_workaround
+#include <windows.h>
```
Are you including this because of `process_t`? If so, you should probably find a way around this because it pollutes the user namespace. I cleaned a bunch of this stuff up in #1445 for that reason. I mean, it won't pollute unless your extension is enabled, but regardless.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/procinfo.h:
```diff
@@ -0,0 +1,142 @@
+void wid_to_top(wid_t wid);
+
+// add a parent window id to window id
+void wid_set_pwid(wid_t wid, wid_t pwid);
```
The comments are nice, but really, the extra spacing, not so consistent with the rest of the headers elsewhere. Do we really want to do this?

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/xlib/procinfo.cpp:
```diff
@@ -0,0 +1,145 @@
+}
+
+process_t pid_from_wid(wid_t wid) {
+  XSetErrorHandler(XErrorHandlerImpl);
+  XSetIOErrorHandler(XIOErrorHandlerImpl);
```
You seem to repeat these same two lines quite a bit, could possibly extract a helper.

time-killer-games commented:
@RobertBColton Thanks for the review! I'll address everything you pointed out in several hours, right now I'm trying to implement that one missing function there is on Mac.

time-killer-games commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/MacOSX/Cocoa/procinfo.mm:
```diff
@@ -0,0 +1,157 @@
+  }
+}
+
+void cocoa_wid_set_pwid(CGWindowID wid, CGWindowID pwid) {
```
@RobertBColton I'm especially proud of myself for figuring out a way to do this on macOS. Basically, it allows any window regardless of what process it belongs to become like a modal dialog in that it will always stay on top of a given parent window, provided the parent window belongs to the current app. This is useful for a number of reasons, and example is you could have a game that runs in windowed mode which has a main menu with a "more games" button. When that button is selected, you could open an internet browser with your games website or store as the default URL. Then you can get the browser's window id and from that information make it stay on top of your game until the browser is closed, so you won't have to Alt+Tab if you lose focus. (Mac uses a different key combination for task switching, but that's not important). Similarly can be done with opening a readme.txt in the default text editor. Believe it or not, there are a lot of practical uses for a lot of these functions, and I'm only explaining the usefulness of just one of them. This function was a nightmare to figure out on Mac, they make you jump through a lot of hoops to do things that involve working outside of the current process.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/MacOSX/Cocoa/procinfo.mm:
```diff
@@ -0,0 +1,162 @@
+  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
+  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
+  CFIndex windowCount = 0;
+  if ((windowCount = CFArrayGetCount(windowArray))) {
+    for (CFIndex i = 0; i < windowCount; i++) {
+      NSDictionary *windowInfoDictionary =
+      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
+      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
+      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
+      if (level.integerValue < kScreensaverWindowLevel) {
+        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
+        if (wid == windowID.integerValue) {
+          result = true;
+          break;
+        }
+      }
+    }
+  }
```
All of this logic is repeated quite a bit, it's another case for a helper function. Move the loop and all the general logic to a helper that accepts a callback function pointer or `std::function` and then call that inside the loop. Then all of the actual functions can call the helper function and pass a lambda for the logic.

This may seem kind of difficult if you aren't familiar with lambda's already, so I wrote some almost-pseudo code for you. This is for the `cocoa_pid_from_wid` function, but you can adapt the callback to accept more parameters or however many is necessary to implement the body logic of each of these functions. You'll have to fiddle with it a bit to get this to work probably.
```cpp
void loop_helper(std::function<void(NSNumber*)> callback) {
  const CGWindowLevel kScreensaverWindowLevel = CGWindowLevelForKey(kCGScreenSaverWindowLevelKey);
  CFArrayRef windowArray = CGWindowListCopyWindowInfo(kCGWindowListOptionAll, kCGNullWindowID);
  CFIndex windowCount = 0;
  if ((windowCount = CFArrayGetCount(windowArray))) {
    for (CFIndex i = 0; i < windowCount; i++) {
      NSDictionary *windowInfoDictionary =
      (__bridge NSDictionary *)((CFDictionaryRef)CFArrayGetValueAtIndex(windowArray, i));
      NSNumber *ownerPID = (NSNumber *)(windowInfoDictionary[(id)kCGWindowOwnerPID]);
      NSNumber *level = (NSNumber *)(windowInfoDictionary[(id)kCGWindowLayer]);
      if (level.integerValue < kScreensaverWindowLevel) {
        NSNumber *windowID = windowInfoDictionary[(id)kCGWindowNumber];
        if (wid == windowID.integerValue) {
          callback(ownerPID);
          break;
        }
      }
    }
  }
  CFRelease(windowArray);
}

pid_t cocoa_pid_from_wid(CGWindowID wid) {
  pid_t pid;
  loop_helper([&](NSNumber *ownerPID) {
    pid = ownerPID.integerValue;
  });
  return pid;
}
```

time-killer-games replied:
Thanks for the suggestion. Unfortunately I've never used C++ and Objective-C in the same file, (I guess in which case it would be considered Objective-C++) and when I use the compiler flag for it  which is -ObjC++ instead of -ObjC apparently there are some other things that need to be done to use C++ features in the same file, because I've been wondering how to do it for a while now with no success, I get unresolved symbols for some reason. Which is why I have been using C with it in various places instead. I'll have to look more into that because C++ would be a lot more desirable to use for me.

Anyway, I've tried making a helper function to cover all this in the past, but it made the code so very unreadable that I messed up in several places and had to revert the code because I made it no longer work by accident. Objective-C is needlessly verbose. 

I'm tempted to buy that Objective-C to Swift code converter so we have something that is more modern, so it'll last longer, and be much more brief in length. Anyway, thanks!

RobertBColton replied:
That was my mistake, I wasn't thinking about objective C at all. Do you have a similar kind of loop on the other platforms? If so please try what I wrote on Linux/Windows first. We can work through making an objective C version later, it shouldn't be too difficult to find a replacement for `std::function` I don't think. Is it difficult for you to translate what I did into Windows/Linux C++? If so, I can make an attempt to rewrite it for those platforms.

time-killer-games replied:
Don't worry about it, I'll try to make a helper function again for this.

time-killer-games commented:
@RobertBColton I removed all VLA's, reordered the headers to include the C++ headers before the C headers. I added two new functions, cmd_from_pid(pid) and env_from_pid(pid) to get the list of command line paramenters and environment variables from a given application instance's process id. I also got it everything MSVC compliant on Windows. Although, getting environment variables, due to how it was written, only supports 32-bit executables on Windows, which will be patched in the near future, hopefully, if all goes well. The next thing on my to do list is address the Cocoa helper function we talked about. :)

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Linux/procinfo.cpp:
```diff
@@ -0,0 +1,159 @@
+  if (readproc(proc, &proc_info) != 0) {
+    char **args = proc_info.cmdline;
+    size_t cnta = 0;
+    for (; args[cnta] != NULL && strlen(args[cnta]) != 0; cnta++);
+    for (size_t i = 0; i < cnta; i++) {
+      if (string_has_whitespace(args[i])) {
+        cmd += "\"" + string_replace_all(args[i], "\"", "\\\"") + "\"";
+      } else {
+        cmd += args[i];
+      }
+      if (args[i + 1] != NULL) {
+        cmd += " ";
+      }
+    }
+  }
```
This environment parsing logic looks similar to BSD, if this is the case, extract a helper for them please. If Win32/Cocoa environment parsing isn't very different either, also extract helpers for them. I am sure the engine is also already doing this elsewhere in ENIGMA too... so eventually this should ALL go to shared/ instead of being duplicated.

time-killer-games replied:
Thanks for the input, I thought about doing that, but i do feel sometimes adding helper functions for too many aspects of code, especially things that are rather unclear what they actually are supposed to do based on any possible function name, it generally makes code harder to read, but I understand that wanting to not duplicate code is probably more important.

RobertBColton replied:
That's right, it absolutely is important, the only time to not worry about duplicate code is extremely performance critical stuff and edge cases. This is neither of which and the only thing of concern to us is the long term maintainability of ENIGMA when it has few developers.

RobertBColton replied:
In other words, since you've said this isn't going in ENIGMA, the decision you must make as a maintain of your own extension Sam is whether you want to minimize errors and maintenance work for yourself, or you care about how quickly the the environment variables can be parsed. The latter isn't much of a concern because the modern world has optimizing compilers that e.g, know how to inline (literally duplicating a function/code by definition) when it should. However, this is actually the same as the static vs dynamic linking debate. The real decision you are making is the classic tradeoff between time and space.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/procinfo.h:
```diff
@@ -0,0 +1,111 @@
+
+namespace enigma_user {
+
+using procinfo::process_execute;
+using procinfo::process_execute_async;
+using procinfo::process_previous;
+using procinfo::process_evaluate;
+using procinfo::process_clear_pid;
+using procinfo::process_clear_out;
+using procinfo::pid_from_self;
+using procinfo::ppid_from_self;
+using procinfo::path_from_pid;
+using procinfo::dir_from_pid;
+using procinfo::name_from_pid;
+using procinfo::cmd_from_pid;
+using procinfo::env_from_pid;
+using procinfo::pid_exists;
+using procinfo::wid_exists;
+using procinfo::pid_kill;
+using procinfo::window_from_wid;
+using procinfo::wid_from_window;
+using procinfo::pid_from_wid;
+using procinfo::pids_enum;
+using procinfo::ppid_from_pid;
+using procinfo::pids_from_ppid;
+using procinfo::wids_from_pid;
+using procinfo::wid_from_top;
+using procinfo::pid_from_top;
+using procinfo::wid_to_top;
+using procinfo::wid_set_pwid;
+using procinfo::echo;
```
Who told you to keep writing these long using chains?

time-killer-games replied:
lol this was done by accident because it was copied over from my own repository, this update. Taking care of it as we speak.

RobertBColton commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Win32/procinfo.cpp:
```diff
@@ -0,0 +1,451 @@
+          output.append(buffer.data(), dwAvail);
+        }
+      }
+      MSG msg;
+      if (PeekMessage(&msg, NULL, 0, 0, PM_REMOVE)) {
+        TranslateMessage(&msg);
+        DispatchMessage(&msg);
+      }
```
`enigma::handleEvents()`

time-killer-games replied:
Again, this extension I need to be able to use the source code without being dependent on the rest of enigma; This pull request I am not expecting to actually have merged into master, I just made it for code review purposes and for those who care enough to use it in enigma can download it separately because Josh doesn't want this in enigma by default. Sorry for the confusion, I am also building from this exact code as a GM extension so it needs to not have dependency on enigma's main source.

time-killer-games replied:
But i can copy/past the code from handleEvents if i were allowed to relicense it under MIT which i've done with a lot of enigma's code I've written and i don't want to make the GPL enigma has go to shit.

RobertBColton replied:
Helper

RobertBColton replied:
Oh, I see, you just want an honest code review. Alright, with everything I learned about processes from adding the stop button, I will try to catch any bugs in your code.

time-killer-games replied:
Thank you so much Robert! You, Josh, fundies, and the programming community is generally just very helpful without necessarily expecting any personal gain, and it's a joy to work with.

RobertBColton replied:
For this one, it actually doesn't matter, you should still make a helper. It should be using a helper in your own version in your own repo. If you did that you would only need to delete your helper when bringing it into. And to make that even easier, you could put all of the helpers you create, which you know ENIGMA will, into a single file, so when you update ENIGMA, you'll only have to delete a single file.

time-killer-games replied:
That's actually a really good idea. I could define a macro in my dll build scripts, since those are not downloaded with the rest of the repository, and when do an #ifdef for the macro for including the helper header, that way the helpers will only be built into the GM dll's and not into the enigma game exe's.

time-killer-games commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/procinfo.h:
```diff
@@ -0,0 +1,84 @@
+
+namespace enigma_user {
+
+using namespace procinfo;
```
does this condense it enough? I want them to be two different namespaces in case this code were to be used outside of enigma for whatever reason. I actually haven't tested this but it looks right. I'll try it if you find this an acceptable approach.

time-killer-games replied:
@RobertBColton

RobertBColton replied:
It's fine, I was just commenting because I thought I had seen you do that somewhere else in ENIGMA. Please don't add the big ones to ENIGMA, I find them quite spurious boilerplate and an eyesore.

RobertBColton commented:
I just read your comment that you just want an honest review, so ok I'll try to combine my recently acquired knowledge of processes and leave you some detailed feedback later. You may find the conversation on #2069 interesting as I now think you are relevant. I don't feel that there is a very good process library for C++ out there. There is only a proposal for a std::process API and the Qt/wxWidgets ones don't give good control over process groups. Perhaps this is something you could create Sam, and actually add CI for on your own repo and then submit to the MinGW/MSYS/Arch/Debian maintainers as a new package.

It would certainly make integration and use in ENIGMA as an extension much simpler.

time-killer-games commented:
That's a great idea, and I had actually thought about it. I might as well see if I can make it an official package for FreeBSD and Homebrew too for when Mac is fixed.

I think adding a homebrew package will be the easiest to get done because all it requires is the repository being made into a homebrew tap which means I won't have to go through any kind of review process by official package managers. All the hosting I can do on my own with mac. I'm going to start by contacting the MinGW team.

time-killer-games commented:
@RobertBColton feel free to contribute any features you would like to see in this pull request, or suggest ones and I'll do my best to add it. I just fixed the parsing logic to env_from_pid() and it now correctly handles environment variables with values containing an equals sign, in which case it wraps the value in quotes, making differentiation between the equals sign separating the name from the value and the equals signs inside the value easier for the end user to parse.

RobertBColton commented:
Just take a look at 8ecaaa20aca65358bde0775fa206f39c73cd567e how the Unix stop button was just implemented. It uses a process group to put all the processes that the ENIGMA compiler DLL starts during the build process so that it can signal all of them without signaling itself.

time-killer-games commented on ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Linux/procinfo.cpp:
```diff
@@ -0,0 +1,159 @@
+  return env;
+}
+
+string pids_enum(bool trim_dir, bool trim_empty) {
```
@RobertBColton I haven't touched the code used in this function in weeks, perhaps even more than a month. Any idea why since yesterday, calling pids_enum on Mac went from working yesterday and every day before that to always crashing, starting today? Am I using `new` and `delete` correctly?
